BD-4087: Dynamic APNs gateway management (#9165)
braze-inc/braze-docs: _docs/_developer_guide/platforms/swift/push_notifications/dynamic_apns.md

---
nav_title: Dynamic APNs Gateway Management
article_title: Dynamic APNs Gateway Management
platform: Swift
page_order: 30
description: "This page covers how to leverage dynamic APNs gateway management to eliminate the need for manual gateway selection."
channel:
  - push
---

# Dynamic APNs gateway management for push notifications

> > Dynamic Apple Push Notification Service (APNs) gateway management enhances the reliability and efficiency of iOS push notifications by automatically detecting the correct APNs environment. Previously, you would manually selected APNs environments (development or production) for your push notifications, which sometimes led to incorrect gateway configurations, delivery failures, and `BadDeviceToken` errors.

Reviewer: ```suggestion
> Dynamic Apple Push Notification Service (APNs) gateway management enhances the reliability and efficiency of iOS push notifications by automatically detecting the correct APNs environment. Previously, you would manually select APNs environments (development or production) for your push notifications, which sometimes led to incorrect gateway configurations, delivery failures, and `BadDeviceToken` errors.
```

## Benefits

- **Improved reliability:** Notifications are always delivered to the correct APNs environment, reducing failed deliveries.
- **Simplified configuration:** You no longer need to manually manage APNs gateway settings.

> - **Error resilience:** Invalid or missing gateway values are gracefully handled, ensuring uninterrupted service.

Reviewer: ```suggestion
- **Error resilience:** Invalid or missing gateway values are gracefully handled, providing uninterrupted service.
```

## Prerequisites

Braze supports Dynamic APNs gateway management for push notifications on iOS with the following SDK version requirement:

{% sdk_min_versions swift:10.0.0 %}

## How it works

> When an iOS app integrates with the Braze Swift SDK, it sends device-related data, including `apns_gateway` if available, to the Braze SDK API.

Reviewer 2: The actual name of the iOS setting that the SDK reads from the app's entitlements is [`aps-environment`](https://developer.apple.com/documentation/bundleresources/entitlements/aps-environment). `apns_gateway` is just the name that the SDK sends up to the Braze platform. Not sure if we want to make this distinction here.

> When an iOS app integrates with the Braze Swift SDK, it sends device-related data, including `apns_gateway` if available, to the Braze SDK API.

Author: ```suggestion
When an iOS app integrates with the Braze Swift SDK, it sends device-related data, including [`aps-environment`](https://developer.apple.com/documentation/bundleresources/entitlements/aps-environment) if available, to the Braze SDK API.
```

- Development (dev) APNs environment
- Production (prod) APNs environment

### Dynamic gateway updates

Braze stores the reported gateway value for each device. If a new, valid gateway value is received, Braze updates the stored value automatically.

### Push notification delivery

When Braze sends a push notification:

- If a valid gateway value (dev or prod) is stored for the device, Braze uses it to determine the correct APNs environment.
- If no gateway value is stored, Braze defaults to the APNs environment configured in the **App Settings** page.

## Frequently asked questions

### Why was this feature introduced?

With dynamic APNs gateway management, the correct environment is selected automatically. Previously, you had to manually configure the APNs gateway, which could lead to `BadDeviceToken` errors, token invalidation, and potential APNs rate-limiting issues.

### How does this impact push delivery performance?

This feature improves delivery rates by ensuring push tokens are always routed to the correct APNs environment, avoiding failures caused by misconfigured gateways.

### Can I disable this feature?

> Dynamic APNs Gateway Management is turned on by default and provides reliability improvements. If you have specific use cases that require manual gateway selection, contact Braze Support.

Reviewer: ```suggestion
Dynamic APNs Gateway Management is turned on by default and provides reliability improvements. If you have specific use cases that require manual gateway selection, [contact Braze Support]({{site.baseurl}}/help/support).
```
Feel free to disregard this one!

The `apns_gateway` value indicates whether the app is using the:
- Development (dev) APNs environment
- Production (prod) APNs environment

### Dynamic gateway updates

Braze stores the reported gateway value for each device. If a new, valid gateway value is received, Braze updates the stored value automatically.

### Push notification delivery

When Braze sends a push notification:

- If a valid gateway value (dev or prod) is stored for the device, Braze uses it to determine the correct APNs environment.
- If no gateway value is stored, Braze defaults to the APNs environment configured in the **App Settings** page.

## Frequently asked questions

### Why was this feature introduced?

With dynamic APNs gateway management, the correct environment is selected automatically. Previously, you had to manually configure the APNs gateway, which could lead to `BadDeviceToken` errors, token invalidation, and potential APNs rate-limiting issues.

### How does this impact push delivery performance?

This feature improves delivery rates by ensuring push tokens are always routed to the correct APNs environment, avoiding failures caused by misconfigured gateways.

> 

Author: ```suggestion
```

- **Error resilience:** Invalid or missing gateway values are gracefully handled, ensuring uninterrupted service.

## Prerequisites

Braze supports Dynamic APNs gateway management for push notifications on iOS with the following SDK version requirement:

{% sdk_min_versions swift:10.0.0 %}

## How it works

When an iOS app integrates with the Braze Swift SDK, it sends device-related data, including `apns_gateway` if available, to the Braze SDK API.

The `apns_gateway` value indicates whether the app is using the:
- Development (dev) APNs environment
- Production (prod) APNs environment

### Dynamic gateway updates

Braze stores the reported gateway value for each device. If a new, valid gateway value is received, Braze updates the stored value automatically.

### Push notification delivery

When Braze sends a push notification:

- If a valid gateway value (dev or prod) is stored for the device, Braze uses it to determine the correct APNs environment.
- If no gateway value is stored, Braze defaults to the APNs environment configured in the **App Settings** page.

## Frequently asked questions

### Why was this feature introduced?

> With dynamic APNs gateway management, you can automatically select the correct environment. Previously, you had to manually configure the APNs gateway, which could lead to `BadDeviceToken` errors, token invalidation, and potential APNs rate-limiting issues.

Reviewer 3: Correction: you _don't_ automatically select the correct environment. That's done for you.

> With dynamic APNs gateway management, you can automatically select the correct environment. Previously, you had to manually configure the APNs gateway, which could lead to `BadDeviceToken` errors, token invalidation, and potential APNs rate-limiting issues.

Author: ```suggestion
With dynamic APNs gateway management, the correct environment is selected automatically. Previously, you had to manually configure the APNs gateway, which could lead to `BadDeviceToken` errors, token invalidation, and potential APNs rate-limiting issues.
```

### How does this impact push delivery performance?

This feature improves delivery rates by ensuring push tokens are always routed to the correct APNs environment, avoiding failures caused by misconfigured gateways.

> ### Does this feature impact push notification latency?
> 
> No, the feature does not introduce additional latency. It makes sure that notifications are sent through the appropriate APNs gateway.

Author: ```suggestion
```

---
nav_title: Dynamic APNs Gateway Management
article_title: Dynamic APNs Gateway Management
platform: Swift
page_order: 30
description: "This page covers how to leverage dynamic APNs gateway management to eliminate the need for manual gateway selection."
channel:
  - push
---

# Dynamic APNs gateway management for push notifications

> Dynamic Apple Push Notification Service (APNs) gateway management enhances the reliability and efficiency of iOS push notifications by automatically detecting the correct APNs environment, which eliminates the need for manual gateway selection.

> Previously, you would manually selected APNs environments (development or production) for your push notifications, which sometimes led to incorrect gateway configurations, delivery failures, and `BadDeviceToken` errors. The Braze iOS SDK detects and updates the correct APNs gateway based on device-reported data.

Reviewer 3: We need to say, "Now, the Braze iOS..." Some sort of transition into this sentence.

## Benefits

- **Improved reliability:** Notifications are always delivered to the correct APNs environment, reducing failed deliveries.
- **Simplified Configuration:** Developers and marketers no longer need to manually manage APNs gateway settings.

> - **Error Resilience:** Invalid or missing gateway values are gracefully handled, ensuring uninterrupted service.

Reviewer 3: @exigoking @zijinlibraze Can you confirm the point under "Error Resilience" is true?

## Prerequisites

Braze supports Dynamic APNs gateway management for push notifications on iOS with the following SDK version requirement:

{% sdk_min_versions swift:10.0.0 %}

## How it works

When an iOS app integrates with the Braze Swift SDK, it sends device-related data, including `apns_gateway` if available, to the Braze SDK API.

The `apns_gateway` value indicates whether the app is using the:
- Development (dev) APNs environment
- Production (prod) APNs environment

### Dynamic gateway updates

Braze stores the reported gateway value for each device. If a new, valid gateway value is received, Braze updates the stored value automatically.

### Push notification delivery

When Braze sends a push notification:

- If a valid gateway value (dev or prod) is stored for the device, Braze uses it to determine the correct APNs environment.
- If no gateway value is stored, Braze defaults to the APNs environment configured in the **App Settings** page.

> ## Turning on dynamic APNs gateway management

Reviewer 3: Do we need to include this section at all?

1. Upgrade to the latest Braze iOS SDK to start using this feature.
2. Check your APNs credentials in the **App Settings** page to ensure proper configuration.
3. Monitor push notification delivery rates in the Braze reporting dashboard.

## Frequently asked questions

### Why was this feature introduced?

> Previously, customers had to manually configure the APNs gateway, which led to `BadDeviceToken` errors, token invalidation, and potential APNs rate-limiting issues. Dynamic APNs Gateway Management solves these issues by automatically selecting the correct environment.

Reviewer 3: which could lead?

### How does this impact push delivery performance?

This feature improves delivery rates by ensuring push tokens are always routed to the correct APNs environment, avoiding failures caused by misconfigured gateways.

> ### Does this feature impact push notification latency?

Reviewer 3: @exigoking is this a valid question or can we remove?
